### PR TITLE
Re-revert allowing timeouts for commands and downloads.

### PR DESCRIPTION
**Pull request by reitermarkus**
- [x] Have you followed the guidelines in our [Contributing](https://github.com/Homebrew/brew/blob/HEAD/CONTRIBUTING.md) document?
- [x] Have you checked to ensure there aren't other open [Pull Requests](https://github.com/Homebrew/brew/pulls) for the same change?
- [x] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your changes? [Here's an example](https://github.com/Homebrew/brew/blob/HEAD/Library/Homebrew/test/PATH_spec.rb).
- [x] Have you successfully run `brew style` with your changes locally?
- [x] Have you successfully run `brew typecheck` with your changes locally?
- [x] Have you successfully run `brew tests` with your changes locally?

-----

Reverts https://github.com/Homebrew/brew/pull/10922 and https://github.com/Homebrew/brew/pull/10923.

Adds compatibility layer for `def fetch` with arity 0.

### CODE REVIEW

**Review by MikeMcQuaid — DISMISSED** [review]

File: Library/Homebrew/compat/download_strategy.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+      super()
+    end
+  end
+
+  class << self
+    def method_added(method)
+      if method == :fetch && instance_method(method).arity.zero?
+        odeprecated "`def fetch` in a subclass of #{self}",
+                    "`def fetch(timeout: nil, **options)` and output a warning " \
+                    "when `options` contains new unhandled options"
```
Comment (Bo98) [inline Library/Homebrew/compat/download_strategy.rb:16]: We had more complaints than this, such as `_fetch` and `resolve_url_basename_time_file_size`.

Comment (reitermarkus) [reply Library/Homebrew/compat/download_strategy.rb:16]: These are definitely private APIs. The docs specifically mention `def fetch`, nothing else.

Comment (MikeMcQuaid) [reply Library/Homebrew/compat/download_strategy.rb:16]: Regardless, I think it would be good to handle the `_fetch` case at least.

Comment (reitermarkus) [reply Library/Homebrew/compat/download_strategy.rb]: Added a deprecation when calling `_fetch` without `timeout`.

**Pull request merged**